File: hausz/src/lib/BeleptetoRendszer.ts
import { writable } from "svelte/store";

import { szinkron_keres } from "./alap_fuggvenyek";
import { BelepesStatusz } from "./Tipusok";
import { Uzenet } from "./Tipusok";
import { alap_url } from "./Konstansok";
import { hash_keszites } from "./alap_fuggvenyek";
import { uj_valasz_mutatasa } from "./Uzenet";

let loggedin = false;
let username = "";
let admin = false;
let user_id = 0;

export let közös_loggedin = writable(false);
export let közös_user_id = writable(0);
export let közös_username = writable("");
export let közös_admin = writable(false);
export let közös_statusz_lekerve = writable(false);

közös_loggedin.subscribe((uj_ertek) => { loggedin = uj_ertek; })
közös_username.subscribe((uj_ertek) => { username = uj_ertek; })
közös_admin.subscribe((uj_ertek) => { admin = uj_ertek; })
közös_user_id.subscribe((uj_ertek) => { user_id = uj_ertek; })

export async function kilepes() {
    szinkron_keres("/include/belepteto_rendszer.🦀?logout=igen", "", (uzenet: Uzenet<string>) => {
        if( uzenet.eredmeny == 'ok' ) {
            közös_loggedin.set(false);
            közös_username.set("");
            közös_admin.set(false);
            közös_user_id.set(-1);
            uj_valasz_mutatasa(3000, "sima", "Sikeres kilépés");
        } else {
            uj_valasz_mutatasa(3000, "hiba", uzenet.valasz);
        }
    });
}

export async function belepes(felhasználónév: string, jelszó: string) {
    let post_parameterek_salt_keres = new FormData();
    post_parameterek_salt_keres.append('get_salt', 'yes');
    post_parameterek_salt_keres.append('username', felhasználónév);

    szinkron_keres(`${alap_url}/include/belepteto_rendszer.🦀`, post_parameterek_salt_keres, async (uzenet: {eredmeny: string, salt: string, valasz: string}) => {
        if(uzenet.eredmeny == 'ok') {
            let salt = uzenet.salt;
            let post_parameterek_belepes = new FormData();
            post_parameterek_belepes.append('login', 'yes');
            post_parameterek_belepes.append('username', felhasználónév);

            let jelszo = await hash_keszites(jelszó);
            let saltos_jelszo = await hash_keszites(jelszo + salt);
            post_parameterek_belepes.append('sha256_password', saltos_jelszo);

            szinkron_keres("/include/belepteto_rendszer.🦀", post_parameterek_belepes, (uzenet: Uzenet<string>) => {
                if(uzenet.eredmeny == 'ok') {
                    uj_valasz_mutatasa(3000, "ok", "Sikeres belépés");
                    statusz_lekerese();
                } else {
                    uj_valasz_mutatasa(3000, "hiba", uzenet.valasz);
                }
            });
        } else {
            uj_valasz_mutatasa(3000, "hiba", uzenet.valasz);
        }
    });
}

export function statusz_lekerese() {
    szinkron_keres(`${alap_url}/include/belepteto_rendszer?statusz=1`, "", (uzenet: {
        eredmeny: string;
        session_loggedin: string,
        session_username: string,
        session_admin: string,
        session_user_id: number
    }) => {
        if( uzenet.eredmeny == 'ok' ) {
            közös_loggedin.set(uzenet.session_loggedin == "yes");
            közös_username.set(uzenet.session_username);
            közös_admin.set(uzenet.session_admin == "igen");
            közös_user_id.set(uzenet.session_user_id);
        } else {
            közös_loggedin.set(false);
            közös_username.set("");
            közös_admin.set(false);
            közös_user_id.set(-1);
        }
        közös_statusz_lekerve.set(true);
    });
}

File: webszerver/src/oldalak/belepteto_rendszer.rs
use actix_web::HttpResponse;
use crate::{session::Session, alap_fuggvenyek::{isset, list_key}, backend::lekerdezesek::{új_session_beillesztése, session_törlése}};
use crate::alap_fuggvenyek::exit_error;
use crate::alap_fuggvenyek::exit_ok;

static LOG_PREFIX: &str = "[belepteto] ";

pub fn belepteto_rendszer(post: Vec<(String, String)>, get: Vec<(String, String)>, session: Session) -> HttpResponse {
    if isset("get_salt", post.clone()) {
        let salt_username: String;
        if session.loggedin == "yes".to_string() {
            salt_username = session.username;
        } else {
            salt_username = list_key("username", post);
        }

        let salt = match crate::backend::lekerdezesek::salt_lekerdezese(salt_username.as_str()) {
            Ok(result) => result,
            Err(err) => {
                println!("{}\"error\": \"{}\"", LOG_PREFIX, err);
                return HttpResponse::BadRequest().body(exit_error(format!("\"hiba\": \"Nem sikerült a salt lekérdezése.\"")));
            }
        };

        return HttpResponse::Ok().body(exit_ok(format!("\"salt\": \"{}\"", salt)));
    }

    if isset("login", post.clone()) {
        if !isset("username", post.clone()) {
            return HttpResponse::BadRequest().body(exit_error(format!("\"hiba\": \"Nem adtál meg felhasználónevet.\"")));
        }
        if !isset("sha256_password", post.clone()) {
            return HttpResponse::BadRequest().body(exit_error(format!("\"hiba\": \"Nem adtál meg jelszót.\"")));
        }

        let username = list_key("username", post.clone());
        let password = list_key("sha256_password", post.clone());
        
        let felhasznalo = match crate::backend::lekerdezesek::felhasznalo_lekerdezese(crate::alap_fuggvenyek::FelhasználóAzonosítóAdatok::Felhasználónév(username.clone())) {
            Ok(Some(result)) => result,
            Ok(None) => {
                return HttpResponse::BadRequest().body(exit_error(format!("Hibás felhasználónév vagy jelszó.")));
            },
            Err(err) => {
                println!("{}\"error\": \"{}\"", LOG_PREFIX, err);
                return HttpResponse::BadRequest().body(exit_error(format!("Nem sikerült a salt lekérdezése az adatbázisból.")));
            }
        };

        let password_in_db = match crate::alap_fuggvenyek::get_password_part(crate::alap_fuggvenyek::JelszoReszek::Password, felhasznalo.sha256_jelszó.as_str()) {
            Ok(result) => result,
            Err(err) => {
                println!("{}\"error\": \"{}\"", LOG_PREFIX, err);
                return HttpResponse::BadRequest().body(exit_error(format!("\"hiba\": \"Nem sikerült a jelszó ellenőrzése.\"")));
            }
        };

        if password != password_in_db {
            println!("{}Jelszavak nem egyeznek! ({}) ({})", LOG_PREFIX, password, password_in_db);
            return HttpResponse::BadRequest().body(exit_error(format!("Hibás felhasználónév vagy jelszó.")));
        }

        let új_cookie = crate::backend::session_azonosito_generator::random_új_session_azonosító();

        match új_session_beillesztése(új_cookie.clone(), felhasznalo) {
            Ok(_) => {},
            Err(hiba) => {
                println!("{}Hiba a session beillesztésekor: ({})", LOG_PREFIX, hiba);
                return HttpResponse::InternalServerError().body(exit_error(format!("\"hiba\": \"Belső hiba.\"")));
            }
        }

        let cookie = actix_web::cookie::Cookie::build("hausz_session", új_cookie)
            .path("/")
            .secure(true)
            .http_only(true)
            .same_site(actix_web::cookie::SameSite::Strict)
            .max_age(time::Duration::seconds(259200))
            .finish();

        let mut valasz = HttpResponse::Ok()
            .body(exit_ok(format!("Sikeres belépés.")));

        match valasz.add_cookie(&cookie) {
            Ok(result) => result,
            Err(err) => {
                println!("{}\"error\": \"{}\"", LOG_PREFIX, err);
                return HttpResponse::BadRequest().body(exit_error(format!("\"hiba\": \"Nem sikerült a cookie hozzáadása.\"")));
            }
        };

        return valasz;
    }

    if isset("statusz", get.clone()) {
        if session.loggedin != "yes".to_string() {
            return HttpResponse::Ok().body(exit_error(format!("\"eredmeny\": \"hiba\", \"valasz\":\"Nem vagy belépve\"")));
        }

        return HttpResponse::Ok()
            .body(format!(
                    "{{\"eredmeny\": \"ok\", \"session_loggedin\": \"{}\", \"session_username\": \"{}\", \"session_user_id\": {}, \"session_admin\": \"{}\"}}", 
                    session.loggedin,
                    session.username,
                    session.user_id,
                    session.admin,
                ));
    }

    if isset("logout", get.clone()) {
        match session_törlése(session.cookie, session.user_id) {
            Ok(_) => { return HttpResponse::Ok().body(exit_ok(format!("\"eredmeny\": \"ok\", \"valasz\":\"Sikeres kilépés.\""))); },
            Err(hiba) => {
                println!("{}Hiba a session törlésekor: ({})", LOG_PREFIX, hiba);
                return HttpResponse::InternalServerError().body(exit_error(format!("\"hiba\": \"Belső hiba.\"")));
            }
        }
    }

    println!("{}Ismeretlen szándék: POST=({:?}) GET=({:?})", LOG_PREFIX, post, get);
    HttpResponse::BadRequest().body(exit_error(format!("Ismeretlen szándék")))
}